plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java

```diff
@@ -294,6 +306,211 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+
+        if (groupingSets.size() != 1) {
+            return Optional.empty();
+        }
+
+        PinotTableHandle tableHandle = (PinotTableHandle) handle;
+        Optional<AggregationApplicationResult<ConnectorTableHandle>> distinctCountHllResult = applyDistinctCountHll(session, tableHandle, aggregates, assignments, groupingSets);
+        if (distinctCountHllResult.isPresent()) {
+            return distinctCountHllResult;
+        }
+        if (tableHandle.getQuery().isPresent() &&
+                !(tableHandle.getQuery().get().getAggregateColumns().isEmpty() &&
+                        tableHandle.getQuery().get().getGroupingColumns().isEmpty())) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<ConnectorExpression> projections = ImmutableList.builder();
+        ImmutableList.Builder<Assignment> resultAssignments = ImmutableList.builder();
+        ImmutableList.Builder<AggregationExpression> aggregationExpressions = ImmutableList.builder();
+
+        for (AggregateFunction aggregate : aggregates) {
+            Optional<String> inputExpression = getInputExpression(aggregate, assignments);
+            // Do not push COUNT(<column>) into pinot as it converts to COUNT(*)
+            // This will count null values.
+            if (inputExpression.isEmpty() || (isCountFunction(aggregate) && !isCountStarFunction(aggregate))) {
+                return Optional.empty();
+            }
+            // Distinct aggregations other than distinctcounthll are not supported
+            if (aggregate.isDistinct()) {
+                return Optional.empty();
+            }
+            Optional<String> aggregateFunctionName = getAggregationFunctionName(aggregate);
+            if (aggregateFunctionName.isEmpty()) {
+                return Optional.empty();
+            }
+            String outputColumnName = format("%s(%s)", aggregateFunctionName.get(), inputExpression.get());
+            aggregationExpressions.add(new AggregationExpression(
+                    outputColumnName,
+                    inputExpression.get(),
+                    aggregateFunctionName.get()));
+            PinotColumnHandle newColumn = new PinotColumnHandle(outputColumnName, aggregate.getOutputType());
+            projections.add(new Variable(newColumn.getColumnName(), newColumn.getDataType()));
+            resultAssignments.add(new Assignment(outputColumnName, newColumn, aggregate.getOutputType()));
+        }
+        DynamicTable dynamicTable = new DynamicTable(
+                tableHandle.getTableName(),
+                Optional.empty(),
+                ImmutableList.of(),
+                getOnlyElement(groupingSets).stream()
+                        .map(PinotColumnHandle.class::cast)
+                        .map(PinotColumnHandle::getColumnName)
+                        .collect(toImmutableList()),
+                tableHandle.getQuery().flatMap(DynamicTable::getFilter),
```

Comment:
in `DynamicTable` what are the semantics of filters, aggregations, sorting, etc?
Which is applied first, which is second, etc.?

Current ordering of fields suggest that aggregation comes before filter, so this line looks incorrect -- you are transposing pre-existing filter with an aggregation.

If this is "just" a matter of field ordering in DynamicTable, please update it and add commentary like here:

https://github.com/trinodb/trino/blob/b52bcd71d4a3fc17fcb74eb3aa1bd291b4111128/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcTableHandle.java#L47-L56

When doing that, sure the ordering of fields matches that of constructor arguments, they are equally important!

Reply:
Removed AggregationExpression and now the connector returns the exact pinot data type for aggregations. lmk what you think and thanks for the suggestions!